src/v/kafka/server/handlers/incremental_alter_configs.cc

```diff
@@ -49,7 +49,22 @@ static void parse_and_set_shadow_indexing_mode(
     switch (op) {
     case config_resource_operation::remove:
         simode.op = cluster::incremental_update_operation::remove;
-        simode.value = model::negate_shadow_indexing_flag(enabled_value);
+        switch (enabled_value) {
```

Comment:
This works when only one of `redpanda.remote.read` or `redpanda.remote.write` is specified in the request. That is not always the case though, which leads to unexpected behaviour such as this example:
```
% rpk topic create t
% rpk cluster config set cloud_storage_enable_remote_write=true
% rpk topic alter-config t --delete redpanda.remote.write --delete redpanda.remote.read
% rpk topic describe t
CONFIGS
=======
KEY                                   VALUE       SOURCE
redpanda.remote.read                  false       DEFAULT_CONFIG
redpanda.remote.write                 false       DYNAMIC_TOPIC_CONFIG

% rpk topic alter-config t --delete redpanda.remote.write
% rpk topic alter-config t --delete redpanda.remote.read
% rpk topic describe t
CONFIGS
=======
KEY                                   VALUE       SOURCE
redpanda.remote.read                  false       DEFAULT_CONFIG
redpanda.remote.write                 true        DEFAULT_CONFIG
```

Today I learned that rpk works around redpanda's inability to change `redpanda.remote.read` and `redpanda.remote.write` at the same time and issues separate `IncrementalAlterConfigs` requests separately. But it only does that when both are `--set` and doesn't have this workaround extended to `--delete`. Ref: https://github.com/redpanda-data/redpanda/blob/74223be8ce5ee37c165bcfcb0187018cdb6653ed/src/go/rpk/pkg/cli/topic/config.go#L75-L78

I think the *right* thing to do would be to deprecate `topic_properties::shadow_indexing` and break it out into two separate fields `remote_read` and `remote_write` to allow changing them independently. But that would be a sizable chunk of work (perhaps something to discuss with the storage team). Without that, I think it would be best to (1) extend the rpk workaround to delete's as well and (2) error out when both properties are specified to let third-party tools know that we can't do that change instead of silently doing the wrong update. I think we would need to error out in alter_configs as well, that seems to have the same issue in some edge cases. @Lazin what do you think?

Reply:
Yeah, after running into this bug more and more in adjacent work I think we need to deprecate the `shadow_indexing` property as it exists today and Do The Right (and Simple) thing of keeping `remote.read` and `remote.write` permissions in separate `bool`s.

[I've made a JIRA ticket for this](https://redpandadata.atlassian.net/browse/CORE-7743), but it's unlikely I will get to it within the near future.

Reply:
Closing this PR in favour of https://github.com/redpanda-data/redpanda/pull/23545